amd/src/courseformat/courseindex/section.js

```diff
@@ -64,13 +64,33 @@ export default class Component extends ComponentBase {
                 target = state.section.get(this.course.sectionlist[target.number - 1]);
             }
             const moveDirection = Math.sign(target.number - origin.number);
-            let targetEnd = target;
-            while (this.course.sectionlist.length > targetEnd.number + 1
-                    && state.section.get(this.course.sectionlist[targetEnd.number + 1]).levelsan > origin.levelsan) {
-                targetEnd = state.section.get(this.course.sectionlist[targetEnd.number + 1]);
+            let targetPointer = target;
+            if (moveDirection > 0 && !target.indexcollapsed) {
+                let targetChild = target;
+                let targetPointerNext = target;
+                let collapsedLevel = 100;
+                let broke = false;
+                while (this.course.sectionlist.length > targetChild.number + 1
+                && state.section.get(this.course.sectionlist[targetChild.number + 1]).levelsan > target.levelsan) {
+                    targetChild = state.section.get(this.course.sectionlist[targetChild.number + 1]);
+                    if (collapsedLevel >= targetChild.levelsan) {
+                        collapsedLevel = targetChild.indexcollapsed ? targetChild.levelSan : 100;
+                    }
+                    targetPointer = targetPointerNext;
+                    if (collapsedLevel >= targetChild.levelsan) {
+                        targetPointerNext = targetChild;
+                    }
+                    if (targetChild.levelsan <= origin.levelsan) {
+                        broke = true;
+                        break;
+                    }
+                }
+                if (!broke) {
+                    targetPointer = targetPointerNext;
+                }
             }
             return origin.id != target.id && origin.id != sectionzeroid && target.id != sectionzeroid
```

Comment:
I didn't notice that if validateDropData() returns false, it is called on the parent section.  This greatly reduces the complexity of this function, but also means it needs to return true if a section is validated for drop on itself (to prevent the drop being transferred to its parent), then the drop needs to be ignored in the other functions.
**All the code following let target = this.section; and up to and including the return statement needs to be replaced with just the new return statement**--apparently I can't make a suggestion that covers deleted lines.
```suggestion
            return target.levelsan <= origin.levelsan && origin.id != sectionzeroid && target.id != sectionzeroid;
```